### DISCUSSION TITLE
dns.resolve vs dns.lookup

### DISCUSSION THREAD

**Discussion by gugu** [sindresorhus/got#2333] [Ideas]
The module uses dns.lookup for dns requests. While dns.lookup uses OS-level caches, it is based on synchronous gethostbyname call and can be a problem (https://nodejs.org/api/dns.html#implementation-considerations). Got has its own dns cache, so isn't it better to use dns.resolve instead?
Also, it allows to use dns.Resolver object, which allows more customization

**Comment by sindresorhus** [comment]
Got uses `dns.lookup()` by default because it respects system-level DNS configuration (like `/etc/hosts`), which is important for many use cases. However, you can already customize DNS resolution using the `dnsLookup` option.

If you want to use `dns.resolve()` or a custom resolver, you can pass your own lookup function:

```javascript
import dns from 'dns/promises';

await got('https://example.com', {
	dnsLookup: async (hostname) => {
		const addresses = await dns.resolve4(hostname);
		return {address: addresses[0], family: 4};
	}
});
```

For DNS caching, you can use the [`cacheable-lookup`](https://github.com/szmarczak/cacheable-lookup) package, which works great with Got and gives you full control over DNS resolution.

I think the current approch of using `dns.lookup()` by default but allowing customization is the right balance - it works out of the box for most users while still being flexible for advanced use cases.

**Reply by rtcpw** [reply]
The `dnsLookup` function didn't work for me as an async function. I had to use the old callback style instead.
